src/main/java/fi/aalto/cs/apluscourses/intellij/services/PluginSettings.java

```diff
@@ -0,0 +1,27 @@
+package fi.aalto.cs.apluscourses.intellij.services;
+
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.presentation.MainViewModel;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class PluginSettings implements MainViewModelProvider {
+
+  public static final String COURSE_CONFIGURATION_FILE_PATH = "o1_course_config.json";
+
+  @NotNull
+  private final ConcurrentMap<Project, MainViewModel> mainViewModels = new ConcurrentHashMap<>();
```

Comment:
can you hint me with the case where this map will store `MainViewModels` that are not related to the current `project`?

Reply:
If there are multiple projects open in IntelliJ, each of them have a separate MainViewModel (because each of the projects are mapped to some course separately so there are separate domain model objects too)

Reply:
There actually should be some cleanup: if a project is closed, its `MainViewModel` should be removed from the map, so it can be GC'ed. One possibility would be use `WeakHashMap` again, but I am not sure of the lifetime of `Project` objects. (It is even a bit risky to use them as hash keys in general -- I should study if that is something that the designer of `Project` have thought should be possible. I had this issue in mind but forgot it. Thanks for noting!)

Reply:
So the better way to implement cleanup is probably to listen to "project closed" event or something like that and then remove no-more-needed MainViewModels then. I'll check this.

Reply:
It seems that project can be used as hash key, see [here](https://upsource.jetbrains.com/idea-ce/file/idea-ce-e2d423f42b53eafd5c625b93149a0ebc7e49756c/platform/platform-impl/src/com/intellij/openapi/project/impl/DefaultProject.java?nav=3796:3826:focused&line=157&preview=false). (Some nerd humour there also a few lines above :D)

I consider using a WeakHahMap still too risky and instead, just do manual cleanup when project closes.

Reply:
I meant a bit different thing, but it's good there is synergy!

so the point is, that I don't exactly understand how the projects work in the IJ. For example, I have our "intellij-plugin" opened in two IDEs at the same time (it's the same codebase), like here:
![projects_in_IDE](https://user-images.githubusercontent.com/9991098/74586933-824eb700-4ff5-11ea-96b0-6fecc714453f.png)
does it make them separate projects then?

on the other hand, if I go to the list of projects in ToolBox:
![projects?](https://user-images.githubusercontent.com/9991098/74586941-94c8f080-4ff5-11ea-9066-282a7fb7ac51.png)
then there is only one "intellij-plugin"

so what I was asking is, what `MainViewModels` will be associated with Projects in those cases?
also, in the first picture, are they modules or projects, sorry, help me out of this :laughing: ?

Reply:
My understanding: in one IntelliJ IDEA window there can be only one project open at any time. So the entirety of code you have open in one window = a PROJECT.

The project have 0..n modules. Symbol of a module is that folder with a cyan square (well, that marks the dir that is `$MODULE_DIR$` for that module). To see the list of all the modules in the project either
- see **<PROJECT_ROOT>/.idea/modules.xml**,  OR
- menu File->Project Structure...->Project Settings->Modules

One thing to note is that modules are not nested! **There is no tree structure of modules.** One module's `$MODULE_DIR$` can be inside another module's `$MODULE_DIR$` but that doesn't imply any kind of relationship between the modules. That's just how the files of there modules are located in the file system. The dependencies between the modules inside a project is a separate structure (basically, a directed graph) which is defined in the *.iml files of the modules and which is totally independent of the directory structure.

The situation in your picture (the right window) looks a bit weird. Is it possible that you have loaded **intellij-plugin** and **plugin-importer-exporter** as *modules* to **tmc-intellij** *project*.

![image](https://user-images.githubusercontent.com/6066364/74613735-7cea8d00-5119-11ea-9a53-500ffa9359f1.png)


I suggest that we take a look at this together on Monday.

Finally, I am not sure how IntelliJ works when the same *project* is opened in two windows. AFAIK, normally IntelliJ would not allow that but perhaps there can be cases where that happens (like having running both Ultimate and Community versions).